cpp/dolfinx/mesh/utils.cpp

```diff
@@ -60,48 +60,44 @@ std::vector<double> mesh::h(const Mesh& mesh,
                             const xtl::span<const std::int32_t>& entities,
                             int dim)
 {
-  if (dim != mesh.topology().dim())
-    throw std::runtime_error("Cell size when dim ne tdim  requires updating.");
 
-  if (mesh.topology().cell_type() == CellType::prism and dim == 2)
-    throw std::runtime_error("More work needed for prism cell");
-
-  // Get number of cell vertices
-  const CellType type = cell_entity_type(mesh.topology().cell_type(), dim, 0);
-  const int num_vertices = num_cell_vertices(type);
+  // Get entity vertices in geometry dofs
+  xt::xtensor<std::int32_t, 2> geometry_vertices
+      = entities_to_geometry(mesh, dim, entities, false);
+  const std::size_t num_vertices_per_entity = geometry_vertices.shape(1);
 
   // Get geometry dofmap and dofs
-  const Geometry& geometry = mesh.geometry();
-  const graph::AdjacencyList<std::int32_t>& x_dofs = geometry.dofmap();
-  auto geom_dofs
-      = xt::adapt(geometry.x().data(), geometry.x().size(), xt::no_ownership(),
-                  std::vector({geometry.x().size() / 3, std::size_t(3)}));
-  std::vector<double> h_cells(entities.size(), 0);
-  assert(num_vertices <= 8);
-  xt::xtensor_fixed<double, xt::xshape<8, 3>> points;
-  for (std::size_t e = 0; e < entities.size(); ++e)
+  const mesh::Geometry& geometry = mesh.geometry();
+  const xtl::span<const double> x_g = geometry.x();
+
+  // Lambda function to compute norm
+  auto l2_norm = [](auto p0, auto p1)
   {
-    // Get the coordinates  of the vertices
-    auto dofs = x_dofs.links(entities[e]);
+    double norm = 0;
+    for (std::size_t k = 0; k < 3; ++k)
+      norm += std::pow(p0[k] - p1[k], 2);
+    return std::sqrt(norm);
+  };
 
-    // The below should work, but misbehaves with the Intel icpx compiler
-    // xt::view(points, xt::range(0, num_vertices), xt::all())
-    //     = xt::view(geom_dofs, xt::keep(dofs), xt::all());
-    auto points_view = xt::view(points, xt::range(0, num_vertices), xt::all());
-    points_view.assign(xt::view(geom_dofs, xt::keep(dofs), xt::all()));
+  std::vector<double> h_cells(entities.size(), 0);
 
-    // Get maximum edge length
-    for (int i = 0; i < num_vertices; ++i)
+  if (dim > 0)
+  {
+    for (std::size_t e = 0; e < entities.size(); ++e)
     {
-      for (int j = i + 1; j < num_vertices; ++j)
+      // Get maximum distance between any vertex
+      auto e_vertices = xt::row(geometry_vertices, e);
+      for (std::size_t i = 0; i < num_vertices_per_entity; ++i)
       {
-        auto p0 = xt::row(points, i);
```

Comment:
I don't quite see the point of this change while `e_vertices` is still an xtensor type. It seems a but muddled being a mixture of xtensor std STL.

Reply:
There are no problems with taking row-views og xtensors, as they are contiguous in memory.
Eventually `entities_to_geometry` should return an adjacency-list, such that it can support prisms and pyramids, and that would remove the need for this view.

The main improvements in this PR is that the code now works for higher order geometries, sub entities of cells, and does not do any copying when computing `h`.

Reply:
I think it's confusing to be mixing concepts.